lib/Sema/CheckedCSubst.cpp

```diff
@@ -75,18 +76,254 @@ ExprResult Sema::ActOnTypeApplication(ExprResult TypeFunc, SourceLocation Loc,
 
 }
 
+// Type Instantiation
+RecordDecl* Sema::ActOnRecordTypeApplication(RecordDecl *Base, ArrayRef<TypeArgument> TypeArgs) {
+  assert(Base->isGeneric() && "Base decl must be generic in a type application");
+  assert(Base->getCanonicalDecl() == Base && "Base must be canonical decl");
+ 
+  auto &ctx = Base->getASTContext();
+
+  // Unwrap the type arguments from a 'TypeArgument' to the underlying 'Type *'.
+  auto RawArgs = new (ctx) llvm::SmallVector<const Type *, 4>();
+  for (auto TArg : TypeArgs) {
+    RawArgs->push_back(TArg.typeName.getTypePtr());
+  }
+
+  // If possible, just retrieve the application from the cache.
+  // This is needed not only for performance, but for correctness to handle
+  // recursive references in type applications (e.g. a list which contains a list as a field).
+  if (auto Cached = ctx.getCachedTypeApp(Base, *RawArgs)) {
+    return Cached;   
+  }
+
+  // Notice we pass dummy location arguments, since the type application doesn't exist in user code.
+  RecordDecl* Inst = RecordDecl::Create(ctx, Base->getTagKind(), Base->getDeclContext(), SourceLocation(), SourceLocation(),
+    Base->getIdentifier(), Base->getPreviousDecl(), ArrayRef<TypedefDecl*>(nullptr, (size_t)0) /* TypeParams */, Base, TypeArgs);
```

Comment:
Use: static_cast<size_t>(0)

Reply:
Do we even need the cast?

Reply:
Yes, we need it. Switched to static_cast as suggested